diff -urpN linux-orig/net/socket.c linux-patched/net/socket.c
--- linux-orig/net/socket.c	2006-01-15 00:16:02.000000000 -0600
+++ linux-patched/net/socket.c	2006-03-03 16:53:11.000000000 -0600
@@ -2084,6 +2084,12 @@ void socket_seq_show(struct seq_file *se
 }
 #endif /* CONFIG_PROC_FS */
 
+/* Cluster devices need these, or better: kernel interfaces */
+
+EXPORT_SYMBOL_GPL(sys_connect);
+EXPORT_SYMBOL_GPL(sys_recvmsg);
+EXPORT_SYMBOL_GPL(sys_socket);
+
 /* ABI emulation layers need these two */
 EXPORT_SYMBOL(move_addr_to_kernel);
 EXPORT_SYMBOL(move_addr_to_user);
