diff -urpN linux-orig/drivers/md/Kconfig linux-patched/drivers/md/Kconfig
--- linux-orig/drivers/md/Kconfig	2006-01-15 00:16:02.000000000 -0600
+++ linux-patched/drivers/md/Kconfig	2006-03-03 17:12:50.000000000 -0600
@@ -236,5 +236,15 @@ config DM_MULTIPATH_EMC
 	---help---
 	  Multipath support for EMC CX/AX series hardware.
 
+config DM_CSNAP
+	tristate "Cluster snapshot target support"
+	depends on BLK_DEV_DM && EXPERIMENTAL
+	---help---
+	  This device-mapper target allows you to create a virtual device
+	  that can take snapshots of an underlying device.  This device
+	  can be accessed simultaneously by multiple nodes of a cluster.
+
+	  If unsure, say N.
+
 endmenu
 
diff -urpN linux-orig/drivers/md/Makefile linux-patched/drivers/md/Makefile
--- linux-orig/drivers/md/Makefile	2006-01-15 00:16:02.000000000 -0600
+++ linux-patched/drivers/md/Makefile	2006-03-03 17:12:50.000000000 -0600
@@ -37,6 +37,7 @@ obj-$(CONFIG_DM_MULTIPATH_EMC)	+= dm-emc
 obj-$(CONFIG_DM_SNAPSHOT)	+= dm-snapshot.o
 obj-$(CONFIG_DM_MIRROR)		+= dm-mirror.o
 obj-$(CONFIG_DM_ZERO)		+= dm-zero.o
+obj-$(CONFIG_DM_CSNAP)		+= dm-csnap.o
 
 quiet_cmd_unroll = UNROLL  $@
       cmd_unroll = $(PERL) $(srctree)/$(src)/unroll.pl $(UNROLL) \
