diff -rNwup --exclude entries svn/trunk/exec/main.c pjc/trunk/exec/main.c
--- svn/trunk/exec/main.c	2006-03-27 09:34:33.000000000 +0100
+++ pjc/trunk/exec/main.c	2006-03-27 09:34:52.000000000 +0100
@@ -979,6 +979,7 @@ static void confchg_fn (
 
 static void aisexec_uid_determine (void)
 {
+#if 0// PJC not used anyway...see below
 	struct passwd *passwd;
 
 	passwd = getpwnam(OPENAIS_USER);
@@ -987,10 +988,12 @@ static void aisexec_uid_determine (void)
 		openais_exit_error (AIS_DONE_UID_DETERMINE);
 	}
 	ais_uid = passwd->pw_uid;
+#endif
 }
 
 static void aisexec_gid_determine (void)
 {
+#if 0// PJC not used anyway...see below
 	struct group *group;
 	group = getgrnam (OPENAIS_GROUP);
 	if (group == 0) {
@@ -998,6 +1001,7 @@ static void aisexec_gid_determine (void)
 		openais_exit_error (AIS_DONE_GID_DETERMINE);
 	}
 	gid_valid = group->gr_gid;
+#endif
 }
 
 static void aisexec_priv_drop (void)
@@ -1186,8 +1190,8 @@ int main (int argc, char **argv)
 		
 	objdb->objdb_init ();
 	openais_service_default_objdb_set (objdb);
 	openais_service_link_all (objdb, &openais_config);
-
+#if 0 // PJC
 	res = openais_main_config_read (objdb, &error_string, &openais_config, 1);
 	if (res == -1) {
 		log_printf (LOG_LEVEL_NOTICE, "AIS Executive Service: Copyright (C) 2002-2006 MontaVista Software, Inc and contributors.\n");
@@ -1200,7 +1204,7 @@ int main (int argc, char **argv)
 		log_printf (LOG_LEVEL_ERROR, error_string);
 		openais_exit_error (AIS_DONE_AMFCONFIGREAD);
 	}
-
+#endif
 	if (!openais_config.totem_config.interface_count) {
 		res = totem_config_read (&openais_config.totem_config, &error_string, 1);
 		if (res == -1) {
@@ -1209,7 +1213,7 @@ int main (int argc, char **argv)
 			openais_exit_error (AIS_DONE_MAINCONFIGREAD);
 		}
 	}
-
+#if 0 // PJC
 	res = totem_config_keyread (OPENAIS_CONFDIR "/authkey", &openais_config.totem_config, &error_string);
 	if (res == -1) {
 		log_printf (LOG_LEVEL_ERROR, error_string);
@@ -1221,8 +1225,8 @@ int main (int argc, char **argv)
 		log_printf (LOG_LEVEL_ERROR, error_string);
 		openais_exit_error (AIS_DONE_MAINCONFIGREAD);
 	}
-
-	res = log_setup (&error_string, openais_config.logmode, openais_config.logfile);
+#endif
+	res = log_setup (&error_string, LOG_MODE_SYSLOG, openais_config.logfile);
 	if (res == -1) {
 		log_printf (LOG_LEVEL_ERROR, error_string);
 		openais_exit_error (AIS_DONE_LOGSETUP);
diff -rNwup --exclude entries svn/trunk/exec/service.c pjc/trunk/exec/service.c
--- svn/trunk/exec/service.c	2006-03-27 09:34:33.000000000 +0100
+++ pjc/trunk/exec/service.c	2006-03-27 09:37:13.000000000 +0100
@@ -86,6 +86,10 @@ static struct default_service default_se
 	{
 		.name			 = "openais_cpg",
 		.ver			 = 0,
+	},
+	{
+		.name            = "openais_cman",
+	.ver             = 0,
 	}
 };
 
diff -rNwup --exclude entries svn/trunk/exec/totemnet.c pjc/trunk/exec/totemnet.c
--- svn/trunk/exec/totemnet.c	2006-03-27 09:34:33.000000000 +0100
+++ pjc/trunk/exec/totemnet.c	2006-03-27 09:34:52.000000000 +0100
@@ -271,7 +271,7 @@ static int authenticate_and_decrypt (
 	hmac_done (&instance->totemnet_hmac_state, digest_comparison, &len);
 
 	if (memcmp (digest_comparison, header->hash_digest, len) != 0) {
-		instance->totemnet_log_printf (instance->totemnet_log_level_security, "Received message has invalid digest... ignoring.\n");
+//		instance->totemnet_log_printf (instance->totemnet_log_level_security, "Received message has invalid digest... ignoring.\n");
 		res = -1;
 		return (-1);
 	}
