diff -urN -p linux-2.6.11/drivers/block/Kconfig linux/drivers/block/Kconfig
--- linux-2.6.11/drivers/block/Kconfig	2005-03-03 14:08:00.000000000 -0600
+++ linux/drivers/block/Kconfig	2005-03-03 18:15:23.000000000 -0600
@@ -495,6 +495,13 @@ config CDROM_PKTCDVD_WCACHE
 	  this option is dangerous unless the CD-RW media is known good, as we
 	  don't do deferred write error handling yet.
 
+config BLK_DEV_GNBD
+	tristate "Global network block device support"
+	depends on NET
+	---help---
+	  
+	  If unsure, say N.
+
 source "drivers/s390/block/Kconfig"
 
 source "drivers/block/Kconfig.iosched"
diff -urN -p linux-2.6.11/drivers/block/Makefile linux/drivers/block/Makefile
--- linux-2.6.11/drivers/block/Makefile	2005-01-03 17:25:53.000000000 -0600
+++ linux/drivers/block/Makefile	2005-03-03 18:15:23.000000000 -0600
@@ -44,4 +44,4 @@ obj-$(CONFIG_BLK_DEV_CRYPTOLOOP) += cryp
 obj-$(CONFIG_VIODASD)		+= viodasd.o
 obj-$(CONFIG_BLK_DEV_SX8)	+= sx8.o
 obj-$(CONFIG_BLK_DEV_UB)	+= ub.o
-
+obj-$(CONFIG_BLK_DEV_GNBD)	+= gnbd.o
