# Add lock harness to the build system.
diff -urN -p linux-2.6.9/fs/Kconfig linux/fs/Kconfig
--- linux-2.6.9/fs/Kconfig	2004-10-27 02:11:48.029120217 -0500
+++ linux/fs/Kconfig	2004-10-27 02:11:54.814210981 -0500
@@ -1768,6 +1768,14 @@ config AFS_FS
 config RXRPC
 	tristate
 
+config LOCK_HARNESS
+	tristate "GFS Lock Harness"
+	help
+	  The module that connects GFS to the modules that provide
+	  locking for GFS.
+
+	  If you want to use GFS (a cluster filesystem) say Y here.
+
 endmenu
 
 menu "Partition Types"
diff -urN -p linux-2.6.9/fs/Makefile linux/fs/Makefile
--- linux-2.6.9/fs/Makefile	2004-10-27 02:11:48.029120217 -0500
+++ linux/fs/Makefile	2004-10-27 02:11:54.815210700 -0500
@@ -93,3 +93,4 @@ obj-$(CONFIG_AFS_FS)		+= afs/
 obj-$(CONFIG_BEFS_FS)		+= befs/
 obj-$(CONFIG_HOSTFS)		+= hostfs/
 obj-$(CONFIG_HPPFS)		+= hppfs/
+obj-$(CONFIG_LOCK_HARNESS)	+= gfs_locking/
diff -urN -p linux-2.6.9/fs/gfs_locking/Makefile linux/fs/gfs_locking/Makefile
--- linux-2.6.9/fs/gfs_locking/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ linux/fs/gfs_locking/Makefile	2004-10-27 02:11:54.815210700 -0500
@@ -0,0 +1,14 @@
+###############################################################################
+###############################################################################
+##
+##  Copyright (C) 2004 Red Hat, Inc.  All rights reserved.
+##
+##  This copyrighted material is made available to anyone wishing to use,
+##  modify, copy, or redistribute it subject to the terms and conditions
+##  of the GNU General Public License v.2.
+##
+###############################################################################
+###############################################################################
+
+obj-$(CONFIG_LOCK_HARNESS)	+= lock_harness/
+
diff -urN -p linux-2.6.9/fs/gfs_locking/lock_harness/Makefile linux/fs/gfs_locking/lock_harness/Makefile
--- linux-2.6.9/fs/gfs_locking/lock_harness/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ linux/fs/gfs_locking/lock_harness/Makefile	2004-10-27 02:11:54.815210700 -0500
@@ -0,0 +1,16 @@
+###############################################################################
+###############################################################################
+##
+##  Copyright (C) 2004 Red Hat, Inc.  All rights reserved.
+##
+##  This copyrighted material is made available to anyone wishing to use,
+##  modify, copy, or redistribute it subject to the terms and conditions
+##  of the GNU General Public License v.2.
+##
+###############################################################################
+###############################################################################
+
+obj-$(CONFIG_LOCK_HARNESS) += lock_harness.o
+
+lock_harness-y	:= main.o
+
