diff -urN linux-2.6.9/drivers/block/Kconfig linux-2.6.9-patched/drivers/block/Kconfig
--- linux-2.6.9/drivers/block/Kconfig	2004-10-18 16:53:43.000000000 -0500
+++ linux-2.6.9-patched/drivers/block/Kconfig	2004-10-22 13:27:52.938836304 -0500
@@ -356,6 +356,13 @@
 	  your machine, or if you want to have a raid or loopback device
 	  bigger than 2TB.  Otherwise say N.
 
+config BLK_DEV_GNBD
+	tristate "Global network block device support"
+	depends on NET
+	---help---
+	  
+	  If unsure, say N.
+
 source "drivers/s390/block/Kconfig"
 
 endmenu
diff -urN linux-2.6.9/drivers/block/Makefile linux-2.6.9-patched/drivers/block/Makefile
--- linux-2.6.9/drivers/block/Makefile	2004-10-18 16:54:55.000000000 -0500
+++ linux-2.6.9-patched/drivers/block/Makefile	2004-10-22 13:30:48.224188880 -0500
@@ -43,4 +43,4 @@
 obj-$(CONFIG_VIODASD)		+= viodasd.o
 obj-$(CONFIG_BLK_DEV_SX8)	+= sx8.o
 obj-$(CONFIG_BLK_DEV_UB)	+= ub.o
-
+obj-$(CONFIG_BLK_DEV_GNBD)	+= gnbd.o
