# Add lock harness to the build system.
diff -urN -p linux-2.6.8.1/fs/Kconfig linux/fs/Kconfig
--- linux-2.6.8.1/fs/Kconfig	2004-08-23 09:51:41.744876651 -0500
+++ linux/fs/Kconfig	2004-08-23 09:51:49.448087695 -0500
@@ -1748,6 +1748,14 @@ config AFS_FS
 config RXRPC
 	tristate
 
+config LOCK_HARNESS
+	tristate "GFS Lock Harness"
+	help
+	  The module that connects GFS to the modules that provide
+	  locking for GFS.
+
+	  If you want to use GFS (a cluster filesystem) say Y here.
+
 endmenu
 
 menu "Partition Types"
diff -urN -p linux-2.6.8.1/fs/Makefile linux/fs/Makefile
--- linux-2.6.8.1/fs/Makefile	2004-08-23 09:51:41.745876418 -0500
+++ linux/fs/Makefile	2004-08-23 09:51:49.448087695 -0500
@@ -91,3 +91,4 @@ obj-$(CONFIG_JFS_FS)		+= jfs/
 obj-$(CONFIG_XFS_FS)		+= xfs/
 obj-$(CONFIG_AFS_FS)		+= afs/
 obj-$(CONFIG_BEFS_FS)		+= befs/
+obj-$(CONFIG_LOCK_HARNESS)	+= gfs_locking/
diff -urN -p linux-2.6.8.1/fs/gfs_locking/Makefile linux/fs/gfs_locking/Makefile
--- linux-2.6.8.1/fs/gfs_locking/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ linux/fs/gfs_locking/Makefile	2004-08-23 09:51:49.448087695 -0500
@@ -0,0 +1,14 @@
+###############################################################################
+###############################################################################
+##
+##  Copyright (C) 2004 Red Hat, Inc.  All rights reserved.
+##
+##  This copyrighted material is made available to anyone wishing to use,
+##  modify, copy, or redistribute it subject to the terms and conditions
+##  of the GNU General Public License v.2.
+##
+###############################################################################
+###############################################################################
+
+obj-$(CONFIG_LOCK_HARNESS)	+= lock_harness/
+
diff -urN -p linux-2.6.8.1/fs/gfs_locking/lock_harness/Makefile linux/fs/gfs_locking/lock_harness/Makefile
--- linux-2.6.8.1/fs/gfs_locking/lock_harness/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ linux/fs/gfs_locking/lock_harness/Makefile	2004-08-23 09:51:49.448087695 -0500
@@ -0,0 +1,16 @@
+###############################################################################
+###############################################################################
+##
+##  Copyright (C) 2004 Red Hat, Inc.  All rights reserved.
+##
+##  This copyrighted material is made available to anyone wishing to use,
+##  modify, copy, or redistribute it subject to the terms and conditions
+##  of the GNU General Public License v.2.
+##
+###############################################################################
+###############################################################################
+
+obj-$(CONFIG_LOCK_HARNESS) += lock_harness.o
+
+lock_harness-y	:= main.o
+
