diff -ur a/fs/namespace.c b/fs/namespace.c
--- a/fs/namespace.c	2005-10-10 11:04:44.000000000 -0500
+++ b/fs/namespace.c	2005-10-26 17:31:34.422668941 -0500
@@ -43,6 +43,10 @@
 static int hash_mask __read_mostly, hash_bits __read_mostly;
 static kmem_cache_t *mnt_cache; 
 
+/* /sys/fs */
+decl_subsys(fs, NULL, NULL);
+EXPORT_SYMBOL_GPL(fs_subsys);
+
 static inline unsigned long hash(struct vfsmount *mnt, struct dentry *dentry)
 {
 	unsigned long tmp = ((unsigned long) mnt / L1_CACHE_BYTES);
@@ -1453,6 +1457,7 @@
 		i--;
 	} while (i);
 	sysfs_init();
+	subsystem_register(&fs_subsys);
 	init_rootfs();
 	init_mount_tree();
 }
diff -ur a/include/linux/fs.h b/include/linux/fs.h
--- a/include/linux/fs.h	2005-10-10 11:05:03.000000000 -0500
+++ b/include/linux/fs.h	2005-10-26 17:31:12.222061323 -0500
@@ -1246,6 +1246,9 @@
 
 extern int vfs_statfs(struct super_block *, struct kstatfs *);
 
+/* /sys/fs */
+extern struct subsystem fs_subsys;
+
 #define FLOCK_VERIFY_READ  1
 #define FLOCK_VERIFY_WRITE 2
 
Only in b/include/linux: fs.h.orig
