diff -ur a/fs/Kconfig b/fs/Kconfig
--- a/fs/Kconfig	2005-09-01 16:59:28.038753408 +0800
+++ b/fs/Kconfig	2005-09-01 17:09:39.810749928 +0800
@@ -360,6 +360,21 @@
 	          - POSIX ACLs
 	          - readpages / writepages (not user visible)
 
+config GFS2_FS
+	tristate "GFS2 file system support"
+	depends on DLM
+	select FS_POSIX_ACL
+	help
+	A cluster filesystem.
+
+	Allows a cluster of computers to simultaneously use a block device
+	that is shared between them (with FC, iSCSI, NBD, etc...).  GFS reads
+	and writes to the block device like a local filesystem, but also uses
+	a lock module to allow the computers coordinate their I/O so
+	filesystem consistency is maintained.  One of the nifty features of
+	GFS is perfect consistency -- changes made to the filesystem on one
+	machine show up immediately on all other machines in the cluster.
+
 config MINIX_FS
 	tristate "Minix fs support"
 	help
diff -ur a/fs/Makefile b/fs/Makefile
--- a/fs/Makefile	2005-09-01 16:59:28.042752800 +0800
+++ b/fs/Makefile	2005-09-01 17:10:11.211976216 +0800
@@ -105,3 +105,4 @@
 obj-$(CONFIG_OCFS2_FS)		+= ocfs2/
 obj-$(CONFIG_RELAYFS_FS)	+= relayfs/
 obj-$(CONFIG_9P_FS)		+= 9p/
+obj-$(CONFIG_GFS2_FS)		+= gfs2/
