diff -ur a/include/linux/list.h b/include/linux/list.h
--- a/include/linux/list.h	2005-10-10 11:05:03.000000000 -0500
+++ b/include/linux/list.h	2005-10-11 13:11:26.486626846 -0500
@@ -423,6 +423,19 @@
 	     pos = n, n = list_entry(n->member.next, typeof(*n), member))
 
 /**
+ * list_for_each_entry_safe_reverse - iterate backward over list of given type safe against removal of list entry
+ * @pos:	the type * to use as a loop counter.
+ * @n:		another type * to use as temporary storage
+ * @head:	the head for your list.
+ * @member:	the name of the list_struct within the struct.
+ */
+#define list_for_each_entry_safe_reverse(pos, n, head, member)		\
+	for (pos = list_entry((head)->prev, typeof(*pos), member),	\
+		n = list_entry(pos->member.prev, typeof(*pos), member);	\
+	     &pos->member != (head); 					\
+	     pos = n, n = list_entry(n->member.prev, typeof(*n), member))
+
+/**
  * list_for_each_entry_safe_continue -	iterate over list of given type
  *			continuing after existing point safe against removal of list entry
  * @pos:	the type * to use as a loop counter.
