# Add lock_nolock to the build system.
diff -urN -p linux-2.6.9-rc1-mm2/fs/Kconfig linux/fs/Kconfig
--- linux-2.6.9-rc1-mm2/fs/Kconfig	2004-08-31 16:45:14.576496098 -0500
+++ linux/fs/Kconfig	2004-08-31 16:45:14.587493565 -0500
@@ -1798,6 +1798,12 @@ config GFS_FS
 	  GFS is perfect consistency -- changes made to the filesystem on one
 	  machine show up immediately on all other machines in the cluster.
 
+config LOCK_NOLOCK
+	tristate "Lock Nolock"
+	depends on LOCK_HARNESS
+	help
+	  A "fake" lock module that allows GFS to run as a local filesystem.
+
 endmenu
 
 menu "Partition Types"
diff -urN -p linux-2.6.9-rc1-mm2/fs/gfs_locking/Makefile linux/fs/gfs_locking/Makefile
--- linux-2.6.9-rc1-mm2/fs/gfs_locking/Makefile	2004-08-31 16:45:10.882345780 -0500
+++ linux/fs/gfs_locking/Makefile	2004-08-31 16:45:14.587493565 -0500
@@ -11,4 +11,5 @@
 ###############################################################################
 
 obj-$(CONFIG_LOCK_HARNESS)	+= lock_harness/
+obj-$(CONFIG_LOCK_NOLOCK)	+= lock_nolock/
 
diff -urN -p linux-2.6.9-rc1-mm2/fs/gfs_locking/lock_nolock/Makefile linux/fs/gfs_locking/lock_nolock/Makefile
--- linux-2.6.9-rc1-mm2/fs/gfs_locking/lock_nolock/Makefile	1969-12-31 18:00:00.000000000 -0600
+++ linux/fs/gfs_locking/lock_nolock/Makefile	2004-08-31 16:45:14.587493565 -0500
@@ -0,0 +1,16 @@
+###############################################################################
+###############################################################################
+##
+##  Copyright (C) 2004 Red Hat, Inc.  All rights reserved.
+##
+##  This copyrighted material is made available to anyone wishing to use,
+##  modify, copy, or redistribute it subject to the terms and conditions
+##  of the GNU General Public License v.2.
+##
+###############################################################################
+###############################################################################
+
+obj-$(CONFIG_LOCK_NOLOCK) += lock_nolock.o
+
+lock_nolock-y	:= main.o
+
