diff -urN a/drivers/dlm/dlm_internal.h b/drivers/dlm/dlm_internal.h
--- a/drivers/dlm/dlm_internal.h        2005-04-20 10:46:57.321295784 +0800
+++ b/drivers/dlm/dlm_internal.h        2005-04-20 10:46:16.770460448 +0800
@@ -36,7 +36,7 @@
 #include <asm/semaphore.h>
 #include <asm/uaccess.h>
 
-#include "dlm.h"
+#include <linux/dlm.h>
 
 #define DLM_LOCKSPACE_LEN	(64)
 #define DLM_TOSS_SECS		(10)
diff -urN a/drivers/dlm/device.c b/drivers/dlm/device.c
--- a/drivers/dlm/device.c      2005-04-19 13:23:44.000000000 +0800
+++ b/drivers/dlm/device.c      2005-04-20 10:32:12.724774896 +0800
@@ -36,8 +36,8 @@
 #include <linux/idr.h>
 #include <asm/ioctls.h>
 
-#include "dlm.h"
-#include "dlm_device.h"
+#include <linux/dlm.h>
+#include <linux/dlm_device.h>
 #include "lvb_table.h"
 #include "device.h"
